### DISCUSSION TITLE
AttributeError: 'DataModuleClass' object has no attribute 'training_dataset'

### DISCUSSION THREAD

**Discussion by Opss-0** [Lightning-AI/pytorch-lightning#7446] [General]
I am trying to learn `PyTorch Lightning` by writing a very simple `DataModuleClass`. After `prepare_data()`, and `setup()` I am trying to check these functions are working or not. So, I am trying to get the `training` and `validation` dataset from `setup()`. But I am getting an error

    AttributeError: 'DataModuleClass' object has no attribute 'training_dataset'

Code

    def prepare_data(self):
        x = np.random.uniform(0, 10, 10)
        e = np.random.normal(0, self.sigma, len(x))
        
        # Making target or labels
        y = x + e
        
        # Marging x and e for 2 features
        X = np.transpose(np.array([x, e]))
 
        # Converting numpy array to Tensor
        self.x_train_tensor = torch.from_numpy(X).float().to(device)
        self.y_train_tensor = torch.from_numpy(y).float().to(device)
        
        training_dataset = TensorDataset(self.x_train_tensor, self.y_train_tensor)

        self.training_dataset = training_dataset
    
    def setup(self):
        data = self.training_dataset
        self.train_data, self.val_data = random_split(data, [8, 2])
        return self.train_data, self.val_data
        
        
    def train_dataloader(self):
        return DataLoader(self.train_data)

    def val_dataloader(self):
        return DataLoader(self.val_data)
        
    obj = DataModuleClass()
    print(obj.setup())  

Could you tell me why I am getting this error?

**Comment by kyleong** [comment]
I think you need to call the `prepare_data` method explicitly in order to work.
```python
obj = DataModuleClass()
obj.prepare_data()
obj.setup()
```
or you call it inside the `def setup(self)` method.
```python

def setup(self):
    self.prepare_data()
    data = self.training_dataset
    self.train_data, self.val_data = random_split(data, [8, 2])
    return self.train_data, self.val_data
```

`LightningDataModule` has a property `has_prepared_data` which letting you know if ``datamodule.prepare_data()`` has been called or not. Maybe you can make use of that.

To know more, you can look into the [source code](https://pytorch-lightning.readthedocs.io/en/latest/_modules/pytorch_lightning/core/datamodule.html#LightningDataModule)
